servicetalk-http-api/src/main/java/io/servicetalk/http/api/HeaderUtils.java

```diff
@@ -60,16 +64,16 @@
     static final int HASH_CODE_SEED = 0xc2b2ae35;
     public static final BiFunction<? super CharSequence, ? super CharSequence, CharSequence> DEFAULT_HEADER_FILTER =
             (k, v) -> "<filtered>";
+    private static final Set<CharSequence> DEFAULT_DEBUG_HEADER_NAMES = Collections.unmodifiableSet(
+            new HashSet<>(Arrays.asList(CONTENT_TYPE, CONTENT_LENGTH, TRANSFER_ENCODING)));
```

Comment:
I would recommend keeping only `asList` that already returns an unmodifiable list of elements. When you use for-each loop it iterates over elements of the collection. The internal data structure is much simpler in the list (simple array) vs set (nodes with reference to the next element in memory). Therefore, the traversal will be quicker.

The other aspect is that when you use for-each loop with a collection java should create an iterator under the hood to be able to iterate. However, java 6u23 introduces escape analysis that will help to avoid creation of iterator for simple cases (it works for array-list) [1]. Therefore, java will optimize it to a simple memory counter and eliminate iterator allocation.

Here is a difference between unmodifiable set and list:

```java
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import static io.servicetalk.http.api.CharSequences.contentEqualsIgnoreCase;
import static io.servicetalk.http.api.CharSequences.newAsciiString;
import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_LENGTH;
import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_TYPE;
import static io.servicetalk.http.api.HttpHeaderNames.TRANSFER_ENCODING;
import static java.util.Arrays.asList;
import static java.util.Collections.unmodifiableSet;

@Fork(value = 1)
@State(Scope.Benchmark)
@BenchmarkMode(Mode.Throughput)
@Warmup(iterations = 5, time = 3, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 10, time = 3, timeUnit = TimeUnit.SECONDS)
public class ForEachBenchmark {

    private List<CharSequence> list;
    private Set<CharSequence> set;
    private CharSequence header;

    @Setup(Level.Trial)
    public void setup() {
        list = asList(CONTENT_TYPE, CONTENT_LENGTH, TRANSFER_ENCODING);
        set = unmodifiableSet(new HashSet<>(list));
        header = newAsciiString("some-header-name");
    }

    @Benchmark
    public boolean overList() {
        for (CharSequence headerName : list) {
            if (contentEqualsIgnoreCase(header, headerName)) {
                return true;
            }
        }
        return false;
    }

    @Benchmark
    public boolean overSet() {
        for (CharSequence headerName : set) {
            if (contentEqualsIgnoreCase(header, headerName)) {
                return true;
            }
        }
        return false;
    }
}
```

Results:

```
Benchmark                   Mode  Cnt         Score         Error  Units
ForEachBenchmark.overList  thrpt   10  96009696.616 ± 1279098.553  ops/s
ForEachBenchmark.overSet   thrpt   10  25110467.290 ±  294802.130  ops/s
```

Iteration over the list ~4 times faster.

[1] http://psy-lob-saw.blogspot.com/2014/12/the-escape-of-arraylistiterator.html

Reply:
That makes a lot of sense.

One note though, reading from the documentation for `asList` 

> Returns a fixed-size list backed by the specified array. Changes to the returned list "write through" to the array.

therefore, the original list can be changed, I benchmarked `unmodifiableList` but it's about 10-20% slower.

```
Benchmark                          Mode  Cnt         Score          Error  Units
MyBenchmark.overList              thrpt   10  97029293.383 ± 14275818.096  ops/s
MyBenchmark.overUnmodifiableList  thrpt   10  85589843.703 ±   258289.995  ops/s
```

If we should keep only `asList`, how about using normal arrays `CharSequence[]` instead, since it will be faster, but either way, we sacrifice immutability.

Reply:
Because this is a private list, I won't worry too much about immutability. Nobody will try to alter its state if the scope is limited. I think `asList` is fine.